ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1409,20 +1431,50 @@ void instr_execute(in instr_t instr, in uvec3 regs, in mat2 dUV, in vec3 L)
 	uint ndf = instr_getNDF(instr);
 	bool twosided = instr_getTwosided(instr);
 	
+	vec3 scale = decodeRGB19E7(bsdf_data.data[0].xy);
+	vec3 firstParameter;
+	vec3 opacity;
+	if (opcode_isBxDF(opcode))
+	{
+		bool firstParamTexPresence = instr_get1stParamTexPresence(instr);
+		bool opacityTexPresence = instr_getOpacityTexPresence(instr);
+		opacity = textureOrRGB19E7(bsdf_data.data[0].zw, opacityTexPresence, dUV)*scale.z;
+		firstParameter = textureOrRGB19E7(bsdf_data.data[1].xy, firstParamTexPresence, dUV)*scale.x;
+
+		//in case of rough BSDFs, firstParameter.xy is ax,ay
+		if (ndf==NDF_AS)//&& opcode_isRough(opcode)
+			firstParameter.y = textureOrRGB19E7(bsdf_data.data[1].zw, instr_getAlphaVTexPresence(instr), dUV).x*scale.y;
+	}
+
 	//if (twosided)
 	//flip normal
 
-	vec2 a = getAlpha(opcode,ndf,instr,bsdf_data,dUV);
-	float a2 = a.x*a.y;
-	//TODO somehow get alpha here
-	float D = 0.0;
-	if (ndf==NDF_BECKMANN)
-		D = irr_glsl_beckmann(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
-	else if (ndf==NDF_GGX)
-		D = irr_glsl_ggx_trowbridge_reitz(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
-	else if (ndf==NDF_PHONG)
-		D = irr_glsl_blinn_phong(currBSDFParams.isotropic.NdotH, 1.0/a2);
+    float cosFactor;
+    if (opcode_isBSDF(opcode))
+      cosFactor = abs(currBSDFParams.isotropic.NdotL);
+    else
+      cosFactor = max(currBSDFParams.isotropic.NdotL,0.0);
 
+	float a2;
+	float DG = 0.0;
+	if (cosFactor>FLT_MIN && opcode_isRough(opcode))
+	{
+		a2 = firstParameter.x*firstParameter.y;
+		if (ndf==NDF_BECKMANN) {
+			DG = irr_glsl_beckmann(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
+			DG *= irr_glsl_beckmann_smith_height_correlated(currBSDFParams.isotropic.interaction.NdotV_squared, currBSDFParams.isotropic.NdotL_squared, a2);
+			DG /= 4.0*currBSDFParams.isotropic.interaction.NdotV;
+		}
+		else if (ndf==NDF_GGX) {
+			DG = irr_glsl_ggx_trowbridge_reitz(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
+			DG *= irr_glsl_ggx_smith_height_correlated_wo_numerator(a2, currBSDFParams.isotropic.NdotL, currBSDFParams.isotropic.interaction.NdotV);
+			DG *= cosFactor;
+		}
+		else if (ndf==NDF_PHONG) {
+			DG = irr_glsl_blinn_phong(currBSDFParams.isotropic.NdotH, 1.0/a2);
+			DG /= 4.0*currBSDFParams.isotropic.interaction.NdotV;
+		}
```

Comment:
don't bother supporting phong, if the XML loader detects phong swap it out for GGX and issue a warning in the log

Reply:
but its important to support askhim-shirley !

Reply:
done?

Reply:
no, but obviously will be done in final implementation